keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -413,28 +420,178 @@ On Pod Create/Update/Delete:
 
 ### Kube-Proxy
 
-Kube-proxy will be modified to consume EndpointSlice instances besides Endpoints resource. A flag will be added to kube-proxy to toggle the mode. 
-
-```
-Watch: Service, EndpointSlice ==> Manage: iptables, ipvs, etc
-```
-- Merge multiple EndpointSlice into an aggregated list.
-- Reuse the existing processing logic 
-
-### Endpoint Controller (classic)
-
-In order to ensure backward compatibility for external consumer of the core/v1 Endpoints API, the existing K8s endpoint controller will keep running until the API is EOL. The following limitations will apply:
-- Starting from EndpointSlice beta: If # of endpoints in one Endpoints object exceed 1000, generate a warning event to the object. 
-- Starting from EndpointSlice GA: Only include up to 1000 endpoints in one Endpoints Object and throw events.
+Kube-Proxy will support consuming EndpointSlice resources as an alternative to
+Endpoints resources. This will be enabled with the `EndpointSliceProxying`
+feature gate.
+
+This will require minimal changes to proxier implementations, instead updating
+the shared code that transforms Endpoints into a generic data structure to also
+transform EndpointSlices into that same data structure.
+
+### Endpoints Controller (classic)
+
+In order to ensure backwards compatibility for external consumer of the core/v1
+Endpoints API, the existing Endpoints controller will keep running until the
+API is EOL. After EndpointSlices become GA, the Endpoints controller will
+gradually limit functionality.
+
+* Kubernetes 1.20: If the number of endpoints in one Endpoints object exceeds
+  1000, generated a warning event on the Endpoints resource.
+* Kubernetes 1.21: Limit the number of endpoints in one Endpoints object to
+  1000 and generate warning events on Endpoints resource if additional endpoints
+  are dropped.
+
+### EndpointSliceMirroring Controller
+
+In some cases, custom Endpoints resources are created by applications. To ensure
+that these applications will not need to immediately update their
+implementations to also create custom EndpointSlices, a new
+EndpointSliceMirroring controller will be used to mirror custom Endpoints
+resources to corresponding EndpointSlices.
+
+* A new `endpointslicemirroring.kubernetes.io/skip` label will be introduced.
```

Comment:
arbitrary label names! bikeshed time!

I have a slight preference for the prefix being a resource, like `endpointslice.kubernetes.io/...`, and making the part after the slash describe what the label is opting out of. Other label/annotation keys that follow a similar pattern:

* statefulset.kubernetes.io/pod-name
* cronjob.kubernetes.io/instantiate
* endpointslice.kubernetes.io/managed-by

To opt out, are you wanting a true value, a false value, or any value (I have a slight preference for any value, which lets us set to an empty string)?

1. true value means we need a negative sounding label name:
    * `endpointslice.kubernetes.io/skip-mirror=true`
    * `endpointslice.kubernetes.io/no-mirror=true`

2. false value means we need a positive sounding label name:
    * `endpointslice.kubernetes.io/mirror=false`

3. any value means we need a negative sounding label name but don't have the mental overhead of a literal true value (mere presence with any value means we skip)
    * `endpointslice.kubernetes.io/skip-mirror`
    * `endpointslice.kubernetes.io/no-mirror`

Reply:
I think my preference would be for the label value to be meaningful. I think it's confusing for `endpointslice.kubernetes.io/skip-mirror=false` to result in a "true" interpretation. Admittedly we'd have a similar problem with `endpointslice.kubernetes.io/skip-mirror=yes`, but that seems at least slightly less likely. I'd originally thought of this as a controller specific annotation, but because it's something that will be part of the EndpointSlice API, I agree with the `endpointslice` prefix here. So maybe `endpointslice.kubernetes.io/skip-mirror=true`?

Reply:
Updated PR to include this approach.